### ISSUE TITLE
Absurd invariant type error arising in referentially transparent situations

### ISSUE THREAD

**Issue by epgui** [microsoft/pyright#9679]
**Describe the bug**

This is the error I am getting (see code below for full context):

```python
Argument of type "dict[str, str]" cannot be assigned to parameter "query_params" of type "dict[str, str | list[str]] | None" in function "my_minimally_reproducible_function"
  Type "dict[str, str]" is not assignable to type "dict[str, str | list[str]] | None"
    "dict[str, str]" is not assignable to "dict[str, str | list[str]]"
      Type parameter "_VT@dict" is invariant, but "str" is not the same as "str | list[str]"
      Consider switching from "dict" to "Mapping" which is covariant in the value type
    "dict[str, str]" is not assignable to "None"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
(constant) _THING: dict[str, str]
```

While I understand that dict is invariant, the sample code is referentially transparent. The type system is able to infer the correct type when the value is inlined, but infers an incompatible type when the value is bound to a variable, **even when this variable is used in exactly one place** (ie.: it's not like the type checker has to infer type compatibility with competing type hints, there is only one type hint it's checking against).

**Code**

This code does not have any type errors:

```python
from typing import Optional


def my_minimally_reproducible_function(
    query_params: Optional[dict[str, str | list[str]]] = None,
):
    pass


my_minimally_reproducible_function(query_params={"key": "value"})
```

However the following code does:

```python
from typing import Optional

_THING = {"key": "value"}


def my_minimally_reproducible_function(
    query_params: Optional[dict[str, str | list[str]]] = None,
):
    pass


my_minimally_reproducible_function(query_params=_THING)
```

**VS Code extension or command-line**

This is coming up in VSCode / Pylance (not 100% sure which version), and I can confirm that this issue exists with `pyright v1.1.389` when run from the command line.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the Python typing spec, so this isn't a bug.

The type of `_THING` is established at the time it is constructed. Since you haven't provided any type annotation here, its type is inferred according to the standard inference rules documented [here](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions). If you would like pyright to infer a different type for your dictionary expression, you can use a type annotation for the variable you are assigning it to. Pyright will then use bidirectional type inference.

Closing because this isn't a bug. Pyright is working correctly here.
